doc/smart-switch/smart-switch-database-architecture/smart-switch-database-design.md

```diff
@@ -0,0 +1,262 @@
+
+To determine the DPU number, the "featured" daemon should leverage the platform API. However, for the sake of implementation simplicity, the DPU number is extracted directly from the platform_env.conf file firstly.
+
+``` shell
+cat /usr/share/sonic/device/$PLATFORM/platform_env.conf
+NUM_DPU=2
+```
+
+To align with the established multi-ASIC design in SONiC, a new field, `"has_per_dpu_scope": "True"``, is introduced in the database feature table within config_db.json. This field plays a crucial role in ensuring that each DPU database instance is initiated within a dedicated database container. This design approach maintains consistency with SONiC's existing architecture while accommodating the specific requirements of DPU overlay databases.
+
+``` json
+# config_db.json
+
+"database": {
+    "auto_restart": "always_enabled",
+    "delayed": "False",
+    "has_global_scope": "True",
+    "has_per_asic_scope": "True",
+    "has_per_dpu_scope": "True", # New field for DPU database service
+    "high_mem_alert": "disabled",
+    "state": "always_enabled",
+    "support_syslog_rate_limit": "true"
+},
+```
+
+Within the NPU, the management of DPU overlay databases involves specific configurations. Each DPU overlay database instance is bound to the IP address of the midplane bridge (169.254.200.254 by default). The TCP port assignment follows a predictable pattern, with each DPU ID associated with a unique port (6380 + DPU ID). Additionally, the Unix domain socket path is structured as /var/run/redisdpu{DPU_ID}.
+
+Here is an example includes two DPU:
+
+``` json
+# DPU0: /var/run/redisdpu2/sonic-db/database_config.json
+"redis": {
+    "hostname": "169.254.200.254",
+    "port": 6381,
+    "unix_socket_path": "/var/run/redisdpu0/redis.sock",
+    "persistence_for_warm_boot": "yes",
+    "database_type": "dpudb"
+}
+#DPU1: /var/run/redisdpu1/sonic-db/database_config.json
+"redis": {
+    "hostname": "169.254.200.254",
+    "port": 6382,
+    "unix_socket_path": "/var/run/redisdpu1/redis.sock",
+    "persistence_for_warm_boot": "yes",
+    "database_type": "dpudb"
+}
+```
+
+There are four new tables introduction for the DPU overlay database:
+
+``` json
+"DPU_APPL_DB": {
+    "id": 15,
+    "separator": ":",
+    "instance": "redis",
+    "format": "proto"
+},
+"DPU_APPL_STATE_DB": {
+    "id": 16,
+    "separator": "|",
+    "instance": "redis"
+},
+"DPU_STATE_DB": {
+    "id": 17,
+    "separator": "|",
+    "instance": "redis"
+},
+"DPU_COUNTERS_DB": {
+    "id": 18,
+    "separator": ":",
+    "instance": "redis"
+}
+```
+
+#### Database flow
+
+This section outlines critical workflows interacting with the DPU overlay database.
+
+- Update Overlay Objects via GNMI:
```

Comment:
Will be better to change these into headers or indent the content of each section, so it shows better inside of a single list.

Reply:
Done